### PR TITLE
Add : Support outputting the result of 'verify' in YAML format and only outputting the summary of the result of 'verify'

### PR DESCRIPTION
**Pull request by chunriyeqiongsaigao**
1. Added a new command-line parameter `-o`, which stands for `--output`. The value of this parameter represents the format in which the summary of the `verify` results should be outputted. Currently, only YAML format is supported, the value of which is `yaml`. If the command `e2e verify -c e2e.yaml -o yaml` is entered, only the summary in YAML format will be outputted to the CL.

2. Added a new command-line parameter `--summary-only`. If the command `e2e verify -c e2e.yaml --summary-only` is entered, only the `SUMMARY` part of `verify` results will be outputted to the CL.

example：

This is the `cases` section of the `e2e.yaml` config file for `verify`. It provides 4 cases for e2e to `veriy`. And case-1, case-2, and case-4 will pass, while case-3 will fail.

2023-08-17 10-29-01 的屏幕截图

1. A command without `-o yaml` : `./bin/linux/e2e verify -c ../e2e/e2e.yaml `

2023-08-20 15-18-53 的屏幕截图

2. A command with `-o yaml` : `./bin/linux/e2e verify -c ../e2e/e2e.yaml -o yaml`

2023-08-20 15-29-36 的屏幕截图

3. A command without `--summary-only` : `./bin/linux/e2e verify -c ../e2e/e2e.yaml `

2023-08-20 15-18-53 的屏幕截图

4. A command with `--summary-only` : `./bin/linux/e2e verify -c ../e2e/e2e.yaml --summary-only`

2023-08-20 17-20-57 的屏幕截图

### CODE REVIEW

**Comment by wu-sheng** [comment]
What does `y` actually stand for?

**Comment by chunriyeqiongsaigao** [comment]
> What does `y` actually stand for?

In my opinion, the `y` stands for `yaml`.

**Comment by fgksgf** [comment]
@chunriyeqiongsaigao 
I think `--output` and `-o` would be better, just like `kubectl`. And we may need to support more output format in the future. In this case, we can specify `-o yaml` to output the result in YAML.

**Comment by wu-sheng** [comment]
I am a little confused. From what I read from your demo, the `-y` seems to disable the diff, is that relative to the YAML? Somehow?

**Comment by fgksgf** [comment]
> I am a little confused. From what I read from your demo, the `-y` seems to disable the diff, is that relative to the YAML? Somehow?

We need this PR to make e2e (the dev version) output the verify result in yaml format, so that another e2e (the stable version) can use the yaml output to verify if the dev version e2e works well, just like we discussed in [this](https://github.com/apache/skywalking/discussions/9558)

**Comment by wu-sheng** [comment]
OK. This seems more like `summary in YAML` because actually you disabled many other things.

@kezhenxu94 Are you aware of the naming thing? Any suggestion?

**Comment by fgksgf** [comment]
> OK. This seems more like `summary in YAML` because actually you disabled many other things.

Yes

**Comment by chunriyeqiongsaigao** [comment]
> @chunriyeqiongsaigao I think `--output` and `-o` would be better, just like `kubectl`. And we may need to support more output format in the future. In this case, we can specify `-o yaml` to output the result in YAML.

OK, I will edit the code later. Thanks!

**Comment by wu-sheng** [comment]
If we consider more formats, we should consider this in two parts. 
1. Whether output the verify diff or summary.
2. What is the format of the final result.

**Comment by kezhenxu94** [comment]
> OK. This seems more like `summary in YAML` because actually you disabled many other things.
> 
> 
> 
> @kezhenxu94 Are you aware of the naming thing? Any suggestion?

I'm not quite aware of the naming thing. @chunriyeqiongsaigao Can you ping your mentor here. 

From my understanding, if this is about the output format, then the format name might not be a good command option `-y`, instead you might want to add an option to set the output format such as `-o`, and make `yaml` one of the possible value of that option. `-o yaml`

**Comment by wu-sheng** [comment]
`-o/--output yaml` seems good, meanwhile, could we consider `--output-detail Y/N`(or a better name) to disable the raw diff separately?

**Comment by chunriyeqiongsaigao** [comment]
> I'm not quite aware of the naming thing. @chunriyeqiongsaigao Can you ping your mentor here.
OK！My mentor @fgksgf 👍👍👍

**Comment by fgksgf** [comment]
> `-o/--output yaml` seems good, meanwhile, could we consider `--output-detail Y/N`(or a better name) to disable the raw diff separately?

Make sense, I will discuss with the student.

**Comment by wu-sheng** [comment]
Please update PR title and descriptions to fit your latest codes.

**Comment by chunriyeqiongsaigao** [comment]
> Please update PR title and descriptions to fit your latest codes.
Sure, I will update the PR title and descriptions to align with the latest changes in the code. Thank you for the reminder.

**Comment by chunriyeqiongsaigao** [comment]
> Please update PR title and descriptions to fit your latest codes.

Sure, I will update the PR title and descriptions to align with the latest changes in the code. Thank you for the reminder.

**Comment by wu-sheng** [comment]
I think new command style is good. I want to confirm, is the real tests for this tool will come next?

File: commands/verify/verify.go [hunk 1]
```diff
@@ -33,16 +33,19 @@ import (
...
+	expected    string
+	summaryOnly bool
+	printer     output.Printer
 )
 
 func init() {
 	Verify.Flags().StringVarP(&query, "query", "q", "", "the query to get the actual data, the result of the query should in YAML format")
 	Verify.Flags().StringVarP(&actual, "actual", "a", "", "the actual data file, only YAML file format is supported")
 	Verify.Flags().StringVarP(&expected, "expected", "e", "", "the expected data file, only YAML file format is supported")
+	Verify.Flags().StringVarP(&output.Format, "output", "o", "", "output the verify summary in which format")
```
Comment (fgksgf) [inline commands/verify/verify.go:47]: We had better tell users only yaml is supported in the help text

File: commands/verify/verify.go [hunk 2]
```diff
@@ -33,16 +33,19 @@ import (
...
+	Verify.Flags().BoolVarP(&summaryOnly, "summary-only", "", false, "")
```
Comment (fgksgf) [inline commands/verify/verify.go:48]: No help text?

File: internal/config/globalConfig.go [hunk 3]
```diff
@@ -20,14 +20,15 @@ package config
 
 import (
 	"fmt"
+	"github.com/apache/skywalking-infra-e2e/pkg/output"
```
Comment (fgksgf) [inline internal/config/globalConfig.go:23]: sort imports

File: pkg/output/output.go [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+
+import (
+	"fmt"
+
+	"gopkg.in/yaml.v2"
+)
+
+var (
+	Format  string
+	Formats = []string{"yaml"}
```
Comment (fgksgf) [inline pkg/output/output.go:29]: We can use `map[string]struct{}`

File: pkg/output/output.go [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+)
+
+type YamlCaseResult struct {
+	Passed  []string
+	Failed  []string
+	Skipped []string
+}
+
+func FormatIsNotExist() bool {
```
Comment (fgksgf) [inline pkg/output/output.go:38]: Use map, it will be O(1), instead of O(n).

File: pkg/output/printer.go [hunk 6]
```diff
@@ -19,15 +19,15 @@ package output
...
-
 	"github.com/pterm/pterm"
 )
 
 // CaseResult represents the result of a verification case.
 type CaseResult struct {
-	Msg  string
-	Err  error
-	Skip bool
+	CaseName string
```
Comment (fgksgf) [inline pkg/output/printer.go:27]:
```suggestion
	Name string
```

File: pkg/output/printer.go [hunk 7]
```diff
@@ -40,59 +40,64 @@ type Printer interface {
...
+	spinner        *pterm.SpinnerPrinter
+	batchOutput    bool
+	outputInFormat bool
+	summaryOnly    bool
 }
 
 var _ Printer = &printer{}
 
-func NewPrinter(batchOutput bool) Printer {
+func NewPrinter(batchOutput bool, outputInFormat bool, summaryOnly bool) Printer {
```
Comment (fgksgf) [inline pkg/output/printer.go:51]: Le's use the functional options: https://coolshell.cn/articles/21146.html

**Review by fgksgf — COMMENTED** [review]
Basically correct 👍, but need some polish

**Comment by fgksgf** [comment]
> I think new command style is good. I want to confirm, is the real tests for this tool will come next?

Yes, the student has made a POC.

File: pkg/output/output.go (hunk 4 above)
Comment (fgksgf) [reply pkg/output/output.go:29]: FYI: https://geektutu.com/post/hpg-empty-struct.html

File: pkg/output/output.go (hunk 5 above)
Comment (chunriyeqiongsaigao) [reply pkg/output/output.go:38]: Thanks, Learn a lot !

File: pkg/output/output.go [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+
+import (
+	"fmt"
+
+	"gopkg.in/yaml.v2"
+)
+
+var (
+	Format  string
+	Formats = map[string]struct{}{
```
Comment (fgksgf) [inline pkg/output/output.go:29]:
```suggestion
	SupportedFormats = map[string]struct{}{
```

File: pkg/output/output.go [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+
+func HasFormat() bool {
+	_, ok := Formats[Format]
+	return ok
+}
+
+func PrintResult(caseRes []*CaseResult) {
+	if Format == "yaml" {
+		PrintResultInYAML(caseRes)
+	}
```
Comment (fgksgf) [inline pkg/output/output.go:48]: use `switch` statement?

File: pkg/output/output.go [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+}
+
+func PrintResultInYAML(caseRes []*CaseResult) {
```
Comment (fgksgf) [inline pkg/output/output.go:51]: Do we need to export the function?

File: pkg/output/output.go [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+			} else {
+				yamlCaseResult.Failed = append(yamlCaseResult.Failed, cr.Name)
+			}
+		} else {
+			yamlCaseResult.Skipped = append(yamlCaseResult.Skipped, cr.Name)
+		}
+	}
+
+	yamlData, _ := yaml.Marshal(yamlCaseResult)
+	fmt.Print(string(yamlData))
```
Comment (fgksgf) [inline pkg/output/output.go:66]:
```suggestion
	fmt.Println(string(yamlData))
```

File: pkg/output/printer.go [hunk 12]
```diff
@@ -40,59 +41,86 @@ type Printer interface {
...
+	for _, option := range options {
+		option(p)
+	}
+
+	return p
+}
+
+type Option func(*printer)
+
+func WithBatchMod(batchOutput bool) Option {
```
Comment (fgksgf) [inline pkg/output/printer.go:72]:
```suggestion
func WithBatchOutput(batchOutput bool) Option {
```

File: pkg/output/printer.go [hunk 13]
```diff
@@ -40,59 +41,86 @@ type Printer interface {
...
+	}
+}
+
+func WithFormat(outputInFormat bool) Option {
+	return func(p *printer) {
+		p.outputInFormat = outputInFormat
+	}
+}
+
+func WithSummaryOnly(summaryOnly bool) Option {
```
Comment (fgksgf) [inline pkg/output/printer.go:84]: How about merging the two functions:
```go
func WithSummaryOnly(summaryOnly bool, format string) Option
```

File: pkg/output/printer.go [hunk 14]
```diff
@@ -40,59 +41,86 @@ type Printer interface {
...
+	return func(p *printer) {
+		p.summaryOnly = summaryOnly
 	}
 }
 
 func (p *printer) Start(msg ...string) {
-	if p.batchOutput {
+	if p.batchOutput || p.outputInFormat || p.summaryOnly {
```
Comment (fgksgf) [inline pkg/output/printer.go:91]: I think we do not need to check `outputInFormat` ?

**Review by fgksgf — COMMENTED** [review]
some nits

File: pkg/output/output.go (hunk 9 above)
Comment (chunriyeqiongsaigao) [reply pkg/output/output.go:48]: I think 'switch' is better, but if I use 'switch', the go-lint will go wrong. The go-lint suggests me to use 'if' statement.

**Review by fgksgf — APPROVED** [review]
LGTM, well done!

**Comment by wu-sheng** [comment]
@fgksgf After further tests are added, are we planning a new release?

It has been 24 commits included and over a year since the last release. https://github.com/apache/skywalking-infra-e2e/compare/v1.2.0...main

**Comment by fgksgf** [comment]
> @fgksgf After further tests are added, are we planning a new release?
> 
> It has been 24 commits included and over a year since the last release. [v1.2.0...main](https://github.com/apache/skywalking-infra-e2e/compare/v1.2.0...main)

@wu-sheng No problem, i will do this.

**Pull request merged**